pkg/agent/openflow/pipeline.go

```diff
@@ -2331,78 +2273,82 @@ func (c *client) serviceLearnFlow(groupID binding.GroupIDType, svcIP net.IP, svc
+	if serviceType == v1.ServiceTypeNodePort {
 		// If externalTrafficPolicy of NodePort is Cluster, the first packet of NodePort requires SNAT, so nodeLocalExternal
 		// will be false, and ServiceNeedSNATRegMark will be set. If externalTrafficPolicy of NodePort is Local, the first
 		// packet of NodePort doesn't require SNAT, ServiceNeedSNATRegMark won't be set.
 		unionVal := (ToNodePortAddressRegMark.GetValue() << ServiceEPStateField.GetRange().Length()) + EpToSelectRegMark.GetValue()
-		flowBuilder = ServiceLBTable.BuildFlow(priorityNormal).
-			Cookie(c.cookieAllocator.Request(cookie.Service).Raw()).
+		flowBuilder = ServiceLBTable.ofTable.BuildFlow(priorityNormal).
+			Cookie(cookieID).
 			MatchProtocol(protocol).
 			MatchRegFieldWithValue(NodePortUnionField, unionVal).
 			MatchDstPort(svcPort, nil).
 			Action().LoadRegMark(lbResultMark).
 			Action().LoadRegMark(RewriteMACRegMark)
 		if !nodeLocalExternal {
-			flowBuilder = flowBuilder.Action().LoadRegMark(ServiceNeedSNATRegMark)
+			flowBuilder = flowBuilder.Action().LoadRegMark(ToClusterServiceRegMark)
 		}
 	} else {
 		// If Service type is LoadBalancer, as above NodePort.
-		flowBuilder = ServiceLBTable.BuildFlow(priorityNormal).
-			Cookie(c.cookieAllocator.Request(cookie.Service).Raw()).
+		flowBuilder = ServiceLBTable.ofTable.BuildFlow(priorityNormal).
+			Cookie(cookieID).
 			MatchProtocol(protocol).
 			MatchDstPort(svcPort, nil).
 			MatchDstIP(svcIP).
 			MatchRegMark(EpToSelectRegMark).
 			Action().LoadRegMark(lbResultMark).
 			Action().LoadRegMark(RewriteMACRegMark)
-		if svcType == v1.ServiceTypeLoadBalancer && !nodeLocalExternal {
-			flowBuilder = flowBuilder.Action().LoadRegMark(ServiceNeedSNATRegMark)
+		if serviceType == v1.ServiceTypeLoadBalancer && !nodeLocalExternal {
+			flowBuilder = flowBuilder.Action().LoadRegMark(ToClusterServiceRegMark)
 		}
 	}
 	return flowBuilder.
 		Action().LoadToRegField(ServiceGroupIDField, uint32(groupID)).
 		Action().Group(groupID).Done()
 }
 
-// endpointDNATFlow generates the flow which transforms the Service Cluster IP
-// to the Endpoint IP according to the Endpoint selection decision which is stored
-// in regs.
-func (c *client) endpointDNATFlow(endpointIP net.IP, endpointPort uint16, protocol binding.Protocol) binding.Flow {
-	ipProtocol := getIPProtocol(endpointIP)
+// endpointDNATFlow generates the flow which transforms the Service Cluster IP to the Endpoint IP according to the Endpoint
+// selection decision which is stored in regs.
+func (f *featureService) endpointDNATFlow(endpointIP net.IP, endpointPort uint16, protocol binding.Protocol) binding.Flow {
 	unionVal := (EpSelectedRegMark.GetValue() << EndpointPortField.GetRange().Length()) + uint32(endpointPort)
+	flowBuilder := EndpointDNATTable.ofTable.BuildFlow(priorityNormal).
+		MatchProtocol(protocol).
+		Cookie(f.cookieAllocator.Request(f.category).Raw()).
+		MatchRegFieldWithValue(EpUnionField, unionVal)
+	ipProtocol := getIPProtocol(endpointIP)
```

Comment:
ditto: such changes are not necessary, it takes time to review all of them in such a big patch.
I'm not asking to remove them, but hope we can reach a consensus that this kind of changes should be avoided in the future.

Reply:
OK.